V1 full support based on district level data only

e-kotov opened the pull request:
ok, all done, full support for V1 data and much more cool stuff.

Important: to simplify a few things, the v1 data is now downloaded as csv.gz instead of txt.gz. So before testing, please nuke your cache folder or rename all txt.gz files to csv.gz files. txt.gz files should not get in the way as the functions look specifically for csv.gz files, but if you do not remove them, they will just take up storage space.


Please install with
```r
remotes::install_github(
 "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
 build_vignettes = TRUE,
 force = TRUE
)
```


If the above fails for you because of vignettes, do:

```r
Sys.setenv(PKG_BUILD_VIGNETTES=TRUE)
pak::pkg_install(
    "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
    upgrade = TRUE
)
```

Here's an overview:


```mermaid
flowchart TB
    A["For quick analysis of few dates\nwork with raw CSV.gz data"] -->|"spod_get(
    type = 'origin-destination',
    zones = 'districts',
    dates = c(start = '2020-02-14', end = '2020-03-14') )"
    | F["'tbl' object with 'id' for origins and destinations"]
    
    C["Analyse longer periods (several months)\nor even the whole dataset over several years"]
    -->|"spod_convert_for_analysis(
type = 'origin-destination',
    zones = 'districts',
    dates = c(start = '2020-02-14', end = '2021-05-09') )"| D["path to converted data"]
    D -->|"spod_connect_to_converted_data()" | F["'tbl' object with 'id' for origins and destinations"]
    
    F -->|"dplyr functions: select(), filter(), mutate(), group_by(), summarise(), etc..."| G["dplyr::collect()"]
    G --> H["Result: data.frame / tibble"] --> R[spatial data matched by 'id' with aggegated mobility flows]

    X["spatial data with zones"] --> |"spod_get_zones(
    zones = 'districts',
    ver = 1 )"| Y["polygons with zones in sf object\nwith 'id' that match with origins and destinations"] --> R
```


See the new vignette for testing all the functions.


@eugenividal I am very interested in your opinion on how well the concept of working with this virtual duckdb table that you get from `spod_get()` and `spod_connect_to_converted_data()` is explained in the vignette https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/vignettes/v1-2020-2021-mitma-data-codebook.qmd . Also, suggestions for better function names, specifically the ones that are user-facing functions covered in the vignette.

Robinlovelace commented:
Woohoo, thanks @e-kotov! That is an impressive schematic of function and LOT of commits. Promising. I will give this a proper review by Monday evening latest.

Robinlovelace commented:
Quickfire comment seeing the workflow error message:

```
  '::' or ':::' imports not declared from:
    ‘knitr’ ‘od’ ‘odjitter’ ‘osmactive’ ‘remotes’ ‘zonebuilder’
  'library' or 'require' calls not declared from:
    ‘remotes’ ‘tidyverse’ ‘tmap’
```

If you add those to 'Suggests', GitHub actions should become happier.

e-kotov commented:
> If you add those to 'Suggests', GitHub actions should become happier.

not really) https://github.com/Robinlovelace/spanishoddata/actions/runs/10516439464/job/29138644071?pr=52

anyway, it passes locally

e-kotov commented:
@Robinlovelace are we on for the scheduled Monday meeeting?

Robinlovelace commented:
> @Robinlovelace are we on for the scheduled Monday meeeting?

Yes I can make that.

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,330 @@
+
+Key sources for this codebook/cookbook include:
+
+-   [original data collection methodology in Spanish](https://cdn.mitma.gob.es/portal-web-drupal/covid-19/bigdata/mitma_-_estudio_movilidad_covid-19_informe_metodologico_v3.pdf){target="_blank"} + [automatically translated English version of methodology](https://robinlovelace.github.io/spanishoddata/assets/codebooks/mitma_-_estudio_movilidad_covid-19_informe_metodologico_v3_en.pdf){target="_blank ...
```
the translated files are currently in pkgdown/assets/codebooks. they will be available after the pkgdown website is rebuilt

eugenividal commented:
Hi @e-kotov. Thanks for sending this. I wanted to give you my feedback today, but it will take longer. So, if it is ok, I'll review this, like Robin, at the latest on Monday eve. I'll focus on giving you my opinion on this good-looking virtual table explained in the vignette "Codebook and Cookbook for v1", and thinking of better functions' names.

e-kotov commented:
Ok, I made some minor changes and bug fixes. Important note, there are now three new vignettes:

- https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/vignettes/v1-2020-2021-mitma-data-codebook.qmd (was shortened to prevent information overload)
- moved the data conversion from above here: https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/vignettes/converting-the-data-for-faster-analysis.qmd
- and a separate special case vignette to show how to download all and convert all data here: https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/vignettes/download-and-convert-all-data.qmd

I also simplified the README and froze all code execution, at least for now. Also to save energy of GitHub actions. All plots and tables are currently hardcoded in README. Also hidden advanced details.

eugenividal commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+**Getting the data**
+
+To access the data, use the `spod_get_od()` function. In this example we will use a short interval of dates:
```
spod_get_od() or spod_get()? According to the reference vignette spot_get() is to retrieves the origin-destination data and spod_get_od() to load the OD v1 data for specific dates. This is confusing. 

Also, spod_get_od() is not in Figure 1.

e-kotov replied:
Good catch of `spod_get_od()`. This was just a typo. I removed `spod_get_od()` in favour of having just one single function `spod_get()` which has `type` argument that you can set to "od" or other data available in the data set. So here it is just a mistake.

However, regarding the naming. Perhaps it is good to have `spod_get_od()` that simply runs `spod_get(type='od'). That would be easier to type and better for autocomplete... Or perhaps that would be more confusing, I'm not sure.

eugenividal replied:
Not sure either. If you use spot_get_od() you will have to find a spot_get_[something_else]() for the non-od data.

eugenividal commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
````diff
@@ -0,0 +1,228 @@
+# ℹ Use `print(n = ...)` to see more rows
+```
+
+In this example above, we calculated mean hourly flows over the 4 days of the requested period. The full data for all 4 days was never loaded into memory all at once. Rather the available memory of the computer was used up to its maximum limit to make that calculation happen, without ever exceeding the available memory limit. This is done transparantly to the user with the help of [`DuckDB`](http ...
````
A couple of comments:
1) According to Figure 1, this analysis (of four days) could have been conducted with raw CSV.gz. Is that correct? If so, these different approaches depending on the period of analysis should be explained here (referencing Figure 1).
2) Shouldn't we formally cite (add in references) the DuckDB package? I do not know the conventions for citation in these kinds of documents, but probably better to cite the most relevant R packages you use. What do you think?

e-kotov replied:
1. Figure 1 is actually just 1 day and is not connected to this code. I only inserted it go give a better impression of what the od data represents (I think it does that job quite nicely). The code for this plot is actually kind of long and to reproduce it easily, I would need to write another small function to convert the data we get with `spod_get()` into the format required for the plot.

The analysis of several days is actually conducted in this code on raw CSV.gz data. It's just that instead of reading the files one by one, combining them and making the aggregation, we are doing that using DuckDB (transparent to the user), which gives is an added benefit of cleaner data types in columns, additional columns and on-the-fly conversion to English named variables and variable values. So I'm not sure which two different approaches you are talking about. Just to be clear, even when I say "conduct analysis on raw CSV.gz data", I still mean doing that analysis using DuckDB on these CSV.gz files, without creating any actual database file or converting the data into another format. So perhaps it was just a misunderstanding.
2. Yep, we can cite, I will do it. Good idea. However, note that we also have metadata files: https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/codemeta.json and https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/CITATION.cff where all the packages we use in our package are cited.

eugenividal replied:
1. I think you misunderstood this. I meant Figure 1. I believe you are referring to Figure 2 here.
2. Perfect👌. You know more than myself about citing packages. My understanding was that if a specific package is very important for a project it is better to cite it.

e-kotov replied:
1. I did misunderstand you about which figure we are discussing. But my reply is still kind of relevant. So yes, the analysis for four days could be done on raw CSV.gz and it is.

eugenividal commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+***Note**: As long as you use object `od_dist` created with `spod_get_od()` function, it is much quicker to filter the dates by the `year`, `month` and `day` variables, rather than by the `date` variable.* This is because the data for each day is in a separate CSV file located in folders that look like `year=2020/month=2/day=14`. So when filtering by the `date` field, R will have to scan all CSV  ...
+
+## 2.2. Trips per person data {#tpp-data}
```
It took me a while to understand what this subheading meant. I first interpreted it as averages of trips per person. However, the dataset does not show exactly this. The tables in maestra2 show the number of people by a categorization based on number of trips, which can be 0, 1, 2 or +2 by date and zone. Therefore, I would call this section 2.2. People by number of trips or something similar. I mean the relevant or response variable in that dataset for me is the number of people not the number of trips. Does this make sense to you? People by number of trips is still hard to grasp but I think fairer with the content of the content. Do you agree?

e-kotov replied:
Yep, I agree, this is confusing right now. I will think of a better formulation based on your suggestions.

eugenividal reviewed (commented):
Hi @e-kotov,

Please see my comments on the v1-2020-2021-mitma-data-codebook vignette below. I focused on Figure 1 and the names of the functions (covered in the vignette). I also provided feedback on areas that I found confusing. As I gain a better understanding of the datasets, I realize why some aspects are challenging to explain. Turning complex things into simple ones is not easy, but I believe it is important to make the package accessible to more people.

Robinlovelace commented:
Picking up on this now, great to see tests passing!

Robinlovelace commented:
Checked-out with:

```
gh pr checkout 52
```

e-kotov commented:
@eugenividal Great and thorough review. I will address you comments soon. Quick question, have you actually installed the package from this branch and tried to use it as described?

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+## 2.2. Trips per person data {#tpp-data}
+
+The "trips per person" data shows the number of persons by the number of trips they made from each district or municipality.
```
Will do, good catch.

Robinlovelace commented on CITATION.cff:
```diff
@@ -0,0 +1,340 @@
+# --------------------------------------------
```
Good to have this file.

Robinlovelace commented on DESCRIPTION:
```diff
@@ -30,16 +32,19 @@ Imports:
     xml2
 Suggests: 
+    furrr,
+    future,
```
Really needed? Imported by `furrr`.

Robinlovelace replied:
OK see it is needed..

e-kotov replied:
needed for a helper... perhaps i could add the helper functions to `.Rbuildignore` and move them out of the package scope and let them live as just some scripts in a different folder.

Robinlovelace commented on R/connect_to_converted_data.R:
```diff
@@ -0,0 +1,111 @@
+  data_path,
+  target_table_name = NULL,
+  quiet = FALSE,
+  max_mem_gb = 4, # later increase that to be 4GB or perhaps 60% of available RAM, as for analysis the amount of memory that can be used can significanly affect the speed of aggregations.
```
I would say 

```r
  max_mem_gb = max(2, max_mem_system - 4)
```

Could be good here.

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+# Overall approach to working with data
+
+The @fig-overall-flow presents the overall approach to accessing the data in the `spanishoddata` package.
```
1. I have something along these lines in the updated README, we can move the text description of this schematic here too.

2. Have you read the other vignette on data conversion? Do you understand what these functions do and therefor do you have suggestions for better naming from the perspective of how you understand them?

eugenividal replied:
1. 👌
2. I haven't read the vignette on data conversion. I will have a look at this through the week.

eugenividal replied:
@e-kotov aiming to have a look at this tomorrow.

Robinlovelace commented on R/convert_data.R:
```diff
@@ -1,97 +1,341 @@
-#' Convert all downloaded v1 origin-destination data to duckdb
+#' Convert data for faster analysis
```
```suggestion
#' Convert data from plain text to duckdb or parquet format
```

Robinlovelace commented on R/convert_data.R:
```diff
@@ -1,97 +1,341 @@
+  overwrite = FALSE,
+  data_dir = spod_get_data_dir(),
+  quiet = FALSE,
+  max_mem_gb = 4, # fails on full v1 data with less than 4 GB. 4GB is reasonable.  Eventually I would set this to be either 4 as minumum, or about 60% of available RAM autodetected with `as.numeric(unclass(memuse::Sys.meminfo())[1][['totalram']])` or `unclass(benchmarkme::get_ram())`, whichever is bigger. 
```
See suggestion above.

Robinlovelace commented on R/convert_data.R:
```diff
@@ -1,97 +1,341 @@
+    
+
+    # prepare sql query with `table_to_convert` as source table name
+    sql_save_parquet_query <- dplyr::sql(
```
Neat.

Robinlovelace commented on R/dev_tools.R:
```diff
@@ -0,0 +1,68 @@
+#' @return NULL
+#' 
+#' @keywords internal
+spod_files_sizes <- function(ver = 2) {
```
Nice helper function.

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+-   [original data download page](https://www.transportes.gob.es/ministerio/proyectos-singulares/estudios-de-movilidad-con-big-data/estudios-de-movilidad-anteriores/covid-19/opendata-movilidad){target="_blank"}
+
+-   [original data license](https://opendata-movilidad.mitma.es/LICENCIA%20de%20datos%20abiertos%20del%20MITMA%2020201203.pdf){target="_blank"}, and the (5) [homepage of the open mobility data project of the Ministry of Transport and Sustainable Mobility of Spain](https://www.transportes.gob.es/ministerio/proyectos-singulares/estudio-de-movilidad-con-big-data){target="_blank"}.
```
Ahhh, the acronym of the Ministry is so problematic. They changed the name and the website address several times of the past two years) The data license clearly mentions MITMA and urges the data users to cite it as MITMA open data. I agree we can include the new name too. I will rephrase.
By the way, would you say it is still correct to also refer to them as Secretary of State for Transport, Mobility and Urban Agenda (Secretaria de Estado de Transportes, Movilidad y Agenda Urbana)... This is what they have in the methodology docs. Is that the same as Ministry...? I'm confused here.

eugenividal replied:
I know. It is all Pedro Sanchez fault! :) He had some unpopular periods and reshaped some departments to make the government appealing and efficient. This is the way these departments look in the update 18/07/2024.
![Screenshot from 2024-08-26 15-29-16](https://github.com/user-attachments/assets/30a1deb3-383c-490e-b209-0d41d2ff3d29). Perhaps it is worth it to the data provider how they prefer to be citted?  
The Secretary of State for Transport, Mobility and Urban Agenda (Secretaria de Estado de Transportes, Movilidad y Agenda Urbana) is also old. It does not exist in the screenshot above. However, it seems this Secretary started the project.

e-kotov replied:
I would say it makes sense to cite them according to the docs on the website and provide a link to the website. To me that makes sense and would still be traceable to the whichever name they may have in the future because of the link to the website.

eugenividal replied:
@e-kotov Great! It looks fine in the new version.

eugenividal replied:
@e-kotov Sorry, I said it looks fine, but now that I'm rereading the intro I was wondering whether it wouldn't be better if you removed the translation of the citation of the Secretaría de Estado de Transportes, Movilidad y Agenda Urbana in the citation. I do not think it is necessary (you do not translate the ministries) and that would make it easier to read and avoid those awkward brackets within brackets. Or perhaps remove the Spanish?

Robinlovelace commented on R/download_data.R:
```diff
@@ -16,20 +17,20 @@
 #' spod_download_data(
 #'   type = "od", zones = "municip",
-#'   date_regex = "2020032[0-4]"
+#'   dates = "2020032[0-4]"
```
Great that now works.

Robinlovelace replied:
Do we need to update the `@param dates` description?

e-kotov replied:
I believe I did https://github.com/Robinlovelace/spanishoddata/blob/15fbb32e6c4773080e196d02dc4c7bad659b8939/R/internal_utils.R#L5 it's there because it is then inherited by many functions in the pacakge

Robinlovelace commented on R/get.R:
```diff
@@ -200,6 +202,54 @@ spod_available_data_v2 <- function(
   }
 
+  # add known file sizes from cached data
+  file_sizes <- file_sizes <- readr::read_csv(system.file("extdata", "url_file_sizes_v2.txt.gz", package = "spanishoddata"), show_col_types = FALSE)
```
👍

e-kotov replied:
🤦

Robinlovelace commented on R/get_v1_data.R:
```diff
@@ -215,35 +265,159 @@ spod_get_zones_v1 <- function(
+  # here's a function for that
+  unique_separated_ids <- function(column) {
+    # Split the string by semicolon, remove duplicates, and join them back with semicolons
+    sapply(column, function(x) {
```
`sapply` is not type-safe. Consider using `purrr` or `furrr` functions instead.

e-kotov replied:
will do

e-kotov commented:
@eugenividal 
>If you consider it relevant, I can translate these two documents (they are short), so that we can presented them just as our English translations.

On one hand, that would be great, as machine translation is of course not perfect, on the other, it does not seem like a priority to me at the moment. Also, consider that methodology docs for v2 data are MUCH longer, and we would want them to be translated too if you do that for v1. So... Consider this before you commit to translating) 

I see a middle ground in translating the most important bits of methodologies and the ones that you think the automatic translation got very wrong and including them into our vignettes. I think the text in the vignettes is more important, because the vignettes will be pagers on the pkgdown website of the package and will be indexed better by search engines than any PDFs we attach.

Robinlovelace commented:
@e-kotov there seems to be duplication between these two vignettes:

```
ls vignettes/*convert*
vignettes/converting-the-data-for-faster-analysis.qmd  vignettes/download-and-convert-all-data.qmd
```

Also: not a fan of the filenames, how about `download-convert` or just `convert.qmd`: conversion implied download, right?

Robinlovelace commented:
Suggestion: let's have a single `codebook.qmd` doc.

eugenividal commented:
> @eugenividal Great and thorough review. I will address you comments soon. Quick question, have you actually installed the package from this branch and tried to use it as described?

@e-kotov, I am glad you found my comments useful. 
I just checked installing the package from this branch and tried to use it. I found some issues:

1) I had issues with the instructions you gave in the first comment of this PR. See screenshot below:
![Screenshot from 2024-08-26 12-31-19](https://github.com/user-attachments/assets/f5e04fdb-a79d-4fad-8fcb-22f05aaffe15)

2) I had some issues when running the code in the vignette: 
![Screenshot from 2024-08-26 12-28-10](https://github.com/user-attachments/assets/c83e397f-3883-449e-bae0-6ecb9615c1e2)

I hope that helps.

e-kotov commented on vignettes/converting-the-data-for-faster-analysis.qmd:
```diff
@@ -18,9 +18,9 @@ execute:
 `spod_get()` returns objects that are only appropriate for small datasets representing a few days of the national OD datasets. We recommend converting the data into analysis-ready formats (`DuckDB` or `Parquet`), allowing you to work with large datasets on a consumer laptop (with 8-16 GB of memory).
 
-**Converting the data to either `DuckDB` or `Parquet` will improve the computation time up to 100x compared to continuing working with the tables that you get from `spod_get()` function that reads CSV files every time you do anything.**
+**Converting the data to either `DuckDB` or `Parquet` will improve the computation time up to 100x compared with using data frames returned with `spod_get()`, which reads CSV files every time you call it.**
```
Ok, to be fair, it is more like up to 8-10x, not 100x speed improvement of DuckDB database over CSV files.
![Screenshot 2024-08-26 at 13 31 48](https://github.com/user-attachments/assets/4d4109dc-6090-4e6c-ad88-b42ff99f31b9)

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
````diff
@@ -0,0 +1,228 @@
+spanishoddata::spod_codebook(ver = 1)
+```
+
+The mobility data v1 (2020-2021) was originally released by the [Ministry of Transport and Sustainable Mobility (MITMA)](https://www.transportes.gob.es/) [@mitma_data_v1]. The dataset is produced by [Nommon](https://www.nommon.es/){target="_blank"} using the raw data from [Orange España](https://www.orange.es/){target="_blank"}. Even though the raw data is only from one mobile phone operator, the ...
````
1. Yep, there is much more than OD, esp. in the latest releases of v2 data. Perhaps we'll do a package name change. Though it seems to me the OD data will be of most interest for largest audience.
2. Discussed below. Will adjust.
3. Discussed below too, will adjust.

eugenividal replied:
I agree that OD data will attract more interest. Although 'spanishmobilidata' does not sound bad and would be a more accurate and faithful name for the data it contains. MITMS refers to the project as Open Data Movilidad. What do you think about this, @Robinlovelace?

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+The spatial data you get via `spanishoddata` package is downloaded directly from the source, the geometries of polygons are automatically fixed if there are any invalid geometries. The zone identifiers are stored in `id` column. Apart from that `id` column, the the original zones files do not have any metadata. However, as seen above, using the `spanishoddata` package you get many additional colu ...
+
+# 2. Mobility data
```
yes

e-kotov commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+# 2. Mobility data
+
+All mobility data is referenced via `id_origin`, `id_destination`, or other location identifiers (mostly labelled as `id`) with the two sets of zones described above.
```
In the (to be renamed) "trips per person" data, we don't have origins and destinations, just locations with 'id's. That's what I tried to convey here.

eugenividal replied:
@e-kotov Ah, I understand now! Tricky. How about "All mobility data is referenced using identifiers such as 'id_origin' and 'id_destination', or other location identifiers for non-origin destination data, with the two sets of zones described above."?

e-kotov commented:
> Suggestion: let's have a single `codebook.qmd` doc.

For v1 and v2 in one file? I would refrain from that. In the end, the two datasets are very different. I would separate them on purpose to make it clear they are NOT compatible (and the official methodology doc for v2 emphasises, that the two datasets should not be compared directly, e.g. one cannot directly compare trip counts from 2020-2021 (v1) and 2022+ (v2). Also, in v2 data there will be many more data sets that are not available in v1.

e-kotov commented:
> I just tried installing the package from this branch and using it. I found some issues

@eugenividal 

Thanks for that. Before installing the package try installing quarto:
```r
install.packages("quarto")
```

Overall, this is not a big problem, as one can always install without vignettes. When the package is on CRAN, this will not be an issue. This is only relevant for installs from GitHub.


Since you were not able to install the package because of the errors in the first screenshot, you are not using the latest
version and therefore the code from the vignettes will not work.

e-kotov commented:
@eugenividal you can install without vignettes for now:
```r
remotes::install_github(
 "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
 force = TRUE
)
```

Without vignettes the `spod_codebook` will not work, but other functions should.

eugenividal commented:
> Thanks for that. Before installing the package try installing quarto:

@e-kotov After installing quarto and the package I get the same error messages if I run each of the fragments separately. However, I can render v1-2020-2021-mitma-data-codebook.qmd

e-kotov commented:
> I can render v1-2020-2021-mitma-data-codebook.qmd

Rendering works because no code is actually evaluated, the option that turns this off is in the header of the qmd file.

It seems like you still don't have the version from this particular branch.

If you run `spod_get` (yes without the round bracets):

```r
spod_get
```

You should be getting in console:

```r
function (type = c("od", "origin-destination", "os",
    "overnight_stays", "tpp", "trips_per_person"),
    zones = c("districts", "dist", "distr",
        "distritos", "municipalities", "muni",
        "municip", "municipios"), dates = NULL, data_dir = spod_get_data_dir(),
    quiet = FALSE, max_mem_gb = 3, max_n_cpu = parallelly::availableCores() -
        1, max_download_size_gb = 1, duckdb_target = ":memory:",
    temp_path = spod_get_temp_dir())
{
```

and the rest of the function code...

If your first lines of the output are different, you are not on the relevant package version.

In that case try:

```r
remove.packages("spanishoddata")
remotes::install_github(
 "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
 build_vignettes = FALSE,
 force = TRUE
)
```

eugenividal commented:
> On one hand, that would be great, as machine translation is of course not perfect, on the other, it does not seem like a priority to me at the moment. Also, consider that methodology docs for v2 data are MUCH longer, and we would want them to be translated too if you do that for v1. So... Consider this before you commit to translating)

Agree👌 Not a priority, we can consider this later on.

e-kotov commented:
@eugenividal I don't know why, but for some reason you are getting the "old" version of the package from the `main` branch, instead of getting the one at the `v1-full-support-based-on-district-level-data-only` branch.

eugenividal commented:
>@eugenividal I don't know why, but for some reason you are getting the "old" version of the package from the main branch, instead of getting the one at the v1-full-support-based-on-district-level-data-only branch.

Thanks for spotting the reason, I will try to find a way to fix it.

e-kotov commented:
> Thanks for spotting the reason, I will try to find a way to fix it.

Ideally,

```r
remotes::install_github(
 "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
 build_vignettes = FALSE,
 force = TRUE
)
```

should be installing the `v1-full-support-based-on-district-level-data-only` branch, hence the `@v1-full-support-based-on-district-level-data-only`... But why it fails for you, I have no clue :(

Robinlovelace commented:
> > Suggestion: let's have a single `codebook.qmd` doc.
> 
> For v1 and v2 in one file? I would refrain from that. In the end, the two datasets are very different. I would separate them on purpose to make it clear they are NOT compatible (and the official methodology doc for v2 emphasises, that the two datasets should not be compared directly, e.g. one cannot directly compare trip counts from 2020-2021 (v1) and 2022+ (v2). Also, in v2 data there will be many more data sets that are not available in v1.

:+1:

Robinlovelace commented:
> > Thanks for spotting the reason, I will try to find a way to fix it.
> 
> Ideally,
> 
> ```r
> remotes::install_github(
>  "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
>  build_vignettes = FALSE,
>  force = TRUE
> )
> ```
> 
> should be installing the `v1-full-support-based-on-district-level-data-only` branch, hence the `@v1-full-support-based-on-district-level-data-only`... But why it fails for you, I have no clue :(

You're missing Quarto, should be:

```
remotes::install_github(
 "Robinlovelace/spanishoddata@v1-full-support-based-on-district-level-data-only",
 build_vignettes = FALSE,
 force = TRUE,
  depedencies = TRUE
)
```


That will install the 'deps'.

Robinlovelace commented:
@e-kotov good to go now assuming you're OK with #56. Non-controversial, and tested on V2 data for parquet and duckdb as you'll see. If you *do* want to change the pkg name let's do that as a separate PR, this one is already big!

Robinlovelace commented on vignettes/convert.qmd:
```diff
@@ -29,16 +29,43 @@ library(spanishoddata)
+
+The package will create this directory if it does not exist on the first run of any function that downloads the data.
+
+<details><summary>Setting data directory for advanced users</summary>
```
```suggestion
<details><summary>Setting data directory permanently</summary>
```

Robinlovelace replied:
Not just for "advanced" users: for anyone who wants to have a persistent place to store data across sessions without having to set it manually each time. A large proportion of users.

e-kotov commented:
@Robinlovelace reviewing and fixing a few typos as I go. Will be done soon.
Regarding name change and separate PR - ok. Still thinking about it.

eugenividal commented:
> You're missing Quarto, should be:

Working!!!

e-kotov commented:
@eugenividal just merged latest function name changes (not spod_convert and spod_connect, even if these are still controversial, at least they are short ;) ) So you may want to re-install the package once again using Robin's code with deps=true.

e-kotov commented:
@Robinlovelace Ok, to me this PR seems mostly done. Thanks for improving and combining the vignettes and testing with v2 data. In a few hours I will implement @eugenividal comments on the wording of the codebook and merge this.

Robinlovelace commented:
Great, let me know if you want me to postpone putting it out there but sounds like we're on-track for launch.

Robinlovelace commented:
All on track to merge this by 15:00 UK time (16:00 in Germany I believe) @e-kotov ?

e-kotov commented:
@Robinlovelace yes, cleaned up 99%. Ready. We don't have a logo)

e-kotov commented:
Done. Let's see how the pkgdown website renders and looks live and perhaps will make a few tweaks.

e-kotov commented:
OK, now we are done. We have a working package,

a logo

![](https://github.com/Robinlovelace/spanishoddata/blob/main/man/figures/spanishoddata_biglogo.png?raw=true)

and an opengraph card for social media

![](https://github.com/Robinlovelace/spanishoddata/blob/main/man/figures/card.png?raw=true)

e-kotov commented:
@Robinlovelace can you upload this https://github.com/Robinlovelace/spanishoddata/blob/main/man/figures/card.png?raw=true to Social Preview section of https://github.com/Robinlovelace/spanishoddata/settings ? I cannot access settings.

Robinlovelace commented:
Sure!

eugenividal commented on vignettes/v1-2020-2021-mitma-data-codebook.qmd:
```diff
@@ -0,0 +1,228 @@
+
+# 1. Spatial data with zoning boundaries
+
+The base zoning of the study ([`Distrtics`](#districts)) is made up of census districts in cities and municipalities in the rest of Spain. In areas with low populations, data from nearby areas is aggregated to comply with Data Protection regulations. Based on this base zoning, another zoning at the municipal level ([`Municipalities`](#municipalities)) is generated, including aggregations of munic ...
```
Fantastic!

eugenividal commented:
The logo is super cool @e-kotov. I loved it! 🤩

Just a quick question: is that plot real? I meant does the arrow width represent the volume of trips? I'd have expected most trips in the triangle Madrid - Barcelona - Valencia

e-kotov commented:
@eugenividal it is real. I would have spent the whole day to make it by hand)
I will post the code a bit later, I want to find a good place to put it. Probably should not be in the package itself in the R folder, but in some other folder in the repo. These flows are aggregated and therefore what you see could be more of a visual artifact. There are not between cities, but between provinces.

eugenividal commented:
@e-kotov thanks for your quick reply. That's interesting.

e-kotov commented:
@eugenividal ah, also I did some flows aggregation as there's some bug in flowmapper that I did not have time to deal with
<img width="833" alt="Screenshot 2024-08-28 at 10 22 29" src="https://github.com/user-attachments/assets/6df9fcde-58af-4fd5-a925-67fb9ff9d787">

So the lines in the logo actually combine flows for just 10 nodes and some provinces are aggregated into single point... The image above is with 19 nodes (max allowed by the package)

eugenividal commented:
@e-kotov There are 17 autonomous communities and two autonomous cities (Ceuta and Melilla) in Spain. Perfect for the max allowed in the package.

eugenividal commented on vignettes/convert.qmd:
```diff
@@ -0,0 +1,346 @@
+
+Note that this is a lazily-evaluated in-memory object (note the `:memory:` in the database path). This means that the data is not loaded into memory until you call `collect()` on it. This is useful for quick exploration of the data, but not recommended for large datasets.
+
+## `DuckDB` {#duckdb}
```
Another argument for a restructuring. Why is 2.1 under a heading that says "get in one day" when it actually takes more than one day?

eugenividal commented on vignettes/convert.qmd:
```diff
@@ -0,0 +1,346 @@
+
+-   If you plan to download large amounts of data, we recommend `DuckDB`, as it is one big file and it is easier to update it completely if for example you were to first use only dates for 2020, and then later decided to add more dates from 2021. In this case it would be better to delete the database and create it from scratch.
+
+-   If you only want certain dates, do analysis on them and then add more dates later, `Parquet` may be better, as each day is saved in a separate file, just like the original CSV files. Therefore updating a folder of `Parquet` files is as easy as just creating a new file only for the missing date.
```
Perhaps rephrase: If you only want certain dates, analyse them and add additional dates later.

eugenividal commented on vignettes/convert.qmd:
```diff
@@ -0,0 +1,346 @@
+-   If you only want certain dates, do analysis on them and then add more dates later, `Parquet` may be better, as each day is saved in a separate file, just like the original CSV files. Therefore updating a folder of `Parquet` files is as easy as just creating a new file only for the missing date.
+
+
+In our testing (see @fig-csv-duckdb-parquet-speed), using `DuckDB` database has a significant speed advantage over running the analysis on raw CSV.gz files (`DuckDB` up to 8 times faster), or even on `Parquet` files (`DuckDB` up to 5 times faster). The approaches to working with both formats are exactly the same, just use `{dplyr}` package and its functions for selecting, grouping, filtering and  ...
```
Perhaps rephrase the second sentence: The approaches for working with both formats are the same. Simply use the {dplyr} package and its functions to select, group, filter, and summarize the data.

eugenividal commented on vignettes/convert.qmd:
````diff
@@ -0,0 +1,346 @@
+  dplyr::arrange(date)
+```
+
+To prepare all origin-destination data v1 (2020-2021) for analysis over the whole period of data availability, please follow the steps below:
````
I find this redundant or perhaps place it under the heading name 'Download all available data'

eugenividal reviewed (commented):
Hi @e-kotow, I had a chance to take a look at the vignette "Downloading and converting OD datasets". 
I like the new function names. I think they are effectively balanced between being descriptive and not overly long.
The text is generally clear and easy to understand. However, I believe that restructuring some sections could improve the overall clarity.
This and other minor comments (rephrasing suggestions) are explained below.
I also got some errors when running the code. I could provide you with details later if you need them.

Robinlovelace commented:
Great set of comments @eugenividal. PR welcome, look forward to seeing @e-kotov's comments, your suggestions seem good to me, thank you!

e-kotov commented:
@eugenividal Thanks for a thorough review and feedback. Working on updated version in a new branch here: https://github.com/Robinlovelace/spanishoddata/blob/minor-fixes-ek/vignettes/convert.qmd

eugenividal commented:
@e-kotov No problem. Great! I'll be working on other stuff until Tuesday. Yes, please, send a new PR when it is ready and let me know from Tuesday if you would like me to look at any other aspects.
